morango/syncsession.py

```diff
@@ -173,334 +193,152 @@ def _close_sync_session(self, sync_session):
+        # paginate buffered records so we do not load them all into memory
+        buffered_records = Buffer.objects.filter(transfer_session=self.current_transfer_session)
+        buffered_pages = Paginator(buffered_records, chunk_size)
+        for count in buffered_pages.page_range:
+
+            # serialize and send records to server
+            serialized_recs = BufferSerializer(buffered_pages.page(count).object_list, many=True)
+            self.sync_connection._push_record_chunk(serialized_recs.data)
+
+            # update records_transferred upon successful request
+            self.current_transfer_session.records_transferred += chunk_size
+            self.current_transfer_session.save()
 
     def close_sync_session(self):
 
         # "delete" sync session on server side
         self.sync_connection._close_sync_session(self.sync_session)
 
         # "delete" our own local sync session
-        self.sync_connection.sync_session.active = False
-        self.sync_connection.sync_session.save()
-        self.sync_connection.sync_session = None
-        self = None
+        self.sync_session.active = False
+        self.sync_session.save()
+        self.sync_session = None
 
     def _create_transfer_session(self, push, filter):
 
         # build data for creating transfer session on server side
         data = {
             'id': uuid.uuid4().hex,
-            'filter': filter,
+            'filter': filter.__str__(),
             'push': push,
             'sync_session_id': self.sync_session.id,
-            'local_fsic': DatabaseMaxCounter.calculate_filter_max_counters(filter)
+            'local_fsic': json.dumps(DatabaseMaxCounter.calculate_filter_max_counters(filter))
         }
-        data['start_timestamp'] = timezone.now()
+
         data['last_activity_timestamp'] = timezone.now()
         self.current_transfer_session = TransferSession.objects.create(**data)
+        data.pop('last_activity_timestamp')
 
         # create transfer session on server side
         transfer_resp = self.sync_connection._create_transfer_session(data)
 
-        self.current_transfer_session.remote_fsic = transfer_resp.data.get('local_fsic')
-        if not push:
-            self.current_transfer_session.records_total = transfer_resp.data.get('records_total')
-        self.current_transfer_session.save()
+        self.current_transfer_session.remote_fsic = transfer_resp.json().get('local_fsic')
+        return transfer_resp
 
     def _close_transfer_session(self):
 
         # "delete" transfer session on server side
         self.sync_connection._close_transfer_session(self.current_transfer_session)
 
-        # delete local buffered objects if pushing records
-        if self.current_transfer_session.push:
-            Buffer.objects.filter(transfer_session=self.current_transfer_session).delete()
-            RecordMaxCounterBuffer.objects.filter(transfer_session=self.current_transfer_session).delete()
```

Comment:
Might there not be other times when we would call `_close_transfer_session`? E.g. if someone canceled? And we'd want to clear these out in that case also.